### DISCUSSION TITLE
decorators for parallel constructions

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#223] [Design & Development]
I started this topic in email, but I am now 100% sure it needs to be moved here for the record.   I was looking at the mspass_func_wrapper decorator in mspasspy.util.decorators.py.   I could use some help, but I don't see why we can't create a minor modification of this decorator to handle processes that are actually methods of the data object classes.  This came up when I was working on a basic workflow for our short course notebook that required calls to two methods:   ator (BasicTimeSeries) and free_surface_transformation (Seismogram).  Both would benefit from a modification of the mspass_func_wrapper to handle a method.    Is this possible?   I only have a rudimentary understanding of decorators so I'm not sure of the right syntax but it seems it should be possible because a method is just another function.  There may be some gotcha, however, I'm not aware of.   Can either of you suggest the right syntax?  I can revise mspass_func_wrapper easily enough if I could figure out how to cast the def line for the decorator.

**Comment by pavlis** [comment]
I think the [following url](https://www.tutorialsteacher.com/python/classmethod-decorator) tells how we should do this.  Comments?

**Reply by wangyinz** [reply]
I think the `@classmethod` is for different purpose. It is used to define a method that can be called directly from the class instead of object only.

**Reply by pavlis** [reply]
Ok, thanks for the interpretation.   I'm learning but another example of my limited experience with python.

**Comment by wangyinz** [comment]
Yes, as I replied in the email. We can define functions like:

```
@mspass_func_wrapper
def free_surface_transformation(data, *args, object_history=False, alg_name=None, alg_id=None, dryrun=False, inplace_return=True, **options):
    data.free_surface_transformation(*args)
```

btw, I reviewed the implementation of `mspass_func_wrapper`, and I do think there could be some bugs to be fixed.

**Reply by pavlis** [reply]
Ok, I'll create this set of thin wrappers for all the class methods.   It will be  a good time to fix the decorator problem described in the related discussion thread. (#224)

**Reply by wangyinz** [reply]
@haruming could you please help Gary on the implementation?

**Reply by haruming** [reply]
Sure, what's the name of the branch I will be working on?

**Comment by pavlis** [comment]
I am in the middle of working with the branch and haven’t pushed it to GitHub yet.   Can we do this through the discussion pages?   See what I wrote ad respond there ok?
gary

On Jul 6, 2021, at 12:48 PM, Will Yang ***@***.******@***.***>> wrote:

This message was sent from a non-IU address. Please exercise caution when clicking links or opening attachments from external sources.



Sure, what's the name of the branch I will be working on?

**Comment by pavlis** [comment]
I realized this morning when starting in on the short course tutorials I forgot to do something.   We need to add the mspass_func_decorator to a number of native methods of the data objects.  Examples are:  ator, rtoa, rotate, rotate_to_standard, and several others.  I raise this issue for discussion here for two reasons:
1.  I'm not sure I know how to do this.   These are methods exposed to the python interpreter through pybind11 so how to do this is not totally clear to me.  Can I punt this to one of you?  
2. I suspect no matter what we may need some python wrappers.  If that is so, I had a thought for documentation.  Could we just paste and change the syntax of the doxygen comments on these methods for the docstring of the wrappers?   That way they would appear in the sphynx documentation.   That is a secondary goal if it could be done easier.

**Reply by pavlis** [reply]
Here is the list of methods that need wrappers:

**BasicTimeSeries**
*ator*
*rtoa*

**TimeSeries**
none needed 

**Seismogram**
*rotate*  - three overloaded methods (unclear to me how that should be handled in python)
*rotate_to_standard*
*free_surface_transformation*
*transform* - Looks like you have a wrapper in the pybind11 code.  Side issue is we should be sure it is being tested as it uses some buffer protocol tricks

Note all of these methods operate on the data internal to the object so the wrappers will need to return the altered data.
